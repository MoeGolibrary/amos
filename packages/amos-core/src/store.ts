/*
 * @since 2021-08-02 10:50:14
 * @author junbao <junbao@mymoement.com>
 */

import {
  $amos,
  applyEnhancers,
  createEventCenter,
  Enhancer,
  isAmosObject,
  isObject,
  toType,
  Unsubscribe,
} from 'amos-utils';
import { Box } from './box';
import { withBatch } from './enhancers/withBatch';
import { withCache } from './enhancers/withCache';
import { withConcurrent } from './enhancers/withConcurrent';
import { DevtoolsOptions, withDevtools } from './enhancers/withDevtools';
import { withPreload } from './enhancers/withPreload';
import { Dispatch, Dispatchable, Select, Selectable } from './types';

export type Snapshot = Record<string, unknown>;

export interface StoreOptions {
  /**
   * The preloaded state, which will be loaded to the store with {@link fromJS}.
   * This is effective for {@link withPreload}, which is always loaded.
   */
  preloadedState?: Snapshot;

  /**
   * The store name, use for devtools
   */
  name?: string;

  /**
   * Enable devtools or not, if not set, will auto enable devtools if NODE_ENV
   * is development.
   */
  devtools?: DevtoolsOptions | boolean;
}

export interface Store {
  /**
   * Get the state snapshot or the store
   */
  snapshot: () => Readonly<Snapshot>;

  /**
   * Subscribe the update of the store. The event contains all mutated boxes
   * and its original state.
   *
   * It only will be fired after the root dispatch completed.
   */
  subscribe: (fn: () => void) => Unsubscribe;

  /**
   * Dispatch a dispatchable object to update the state of the store.
   */
  dispatch: Dispatch;

  /**
   * Select state of a box or selector.
   */
  select: Select;
}

export interface EnhanceableStore extends Store {
  /**
   * The mutable state of the store.
   */
  state: Snapshot;

  /**
   * Get initial state for box, default is {@link Box#initialState}.
   * If {@link StoreOptions#preloadedState} is set, will be parsed by
   * {@link import('amos-utils').fromJS}.
   */
  getInitialState: <S>(box: Box<S>) => S;

  /**
   * This function will be called after the store created.
   */
  init: () => void;
}

export type StoreEnhancer = Enhancer<[StoreOptions], EnhanceableStore>;

export function createStore(options: StoreOptions = {}, ...enhancers: StoreEnhancer[]): Store {
  const store = applyEnhancers(
    [withDevtools(), withBatch(), withConcurrent(), withCache(), withPreload(), ...enhancers],
    [options],
    (): EnhanceableStore => {
      const ec = createEventCenter<[]>();
      let isDispatching = false;
      return {
        state: {},
        getInitialState: (box) => box.initialState,
        init: () => void 0,
        snapshot: () => store.state,
        subscribe: ec.subscribe,
        // only accepts Dispatchable here
        dispatch: (_task: any) => {
          const task: Dispatchable = _task;
          if (!isObject(task)) {
            throw new Error(`dispatching non-dispatchable type: ${toType(task)}`);
          }
          const isRoot = !isDispatching;
          isDispatching ||= true;
          try {
            let r: any;
            switch (task[$amos]) {
              case 'action':
                r = task.actor(store.dispatch, store.select, ...task.args);
                break;
              case 'mutation':
                const initialState = store.select(task.box);
                r = task.mutator(initialState);
                store.state[task.box.key] = r;
                break;
              case 'signal':
                r = task.creator(store.select, ...task.args);
                task.factory.dispatch(store.dispatch, store.select, r);
                break;
              default:
                throw new Error(
                  `dispatching non-dispatchable object: ${task[$amos] || toType(task)}`,
                );
            }
            if (isRoot) {
              isDispatching = false;
              ec.dispatch();
            }
            return r;
          } finally {
            if (isRoot && isDispatching) {
              isDispatching = false;
            }
          }
        },
        select: (_selectable: any): any => {
          const selectable: Selectable = _selectable;
          if (isAmosObject<Box>(selectable, 'box')) {
            if (!store.state.hasOwnProperty(selectable.key)) {
              store.state[selectable.key] = store.getInitialState(selectable);
            }
            return store.state[selectable.key];
          }
          return selectable.compute(store.select);
        },
      };
    },
  );
  store.init();
  return {
    snapshot: store.snapshot,
    subscribe: store.subscribe,
    dispatch: store.dispatch,
    select: store.select,
  };
}
