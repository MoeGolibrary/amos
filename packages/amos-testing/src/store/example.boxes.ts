/*
 * @since 2022-01-10 19:51:44
 * @author junbao <junbao@moego.pet>
 *
 * record box
 */

import { createRecordBox } from 'amos-boxes';
import { Record } from 'amos-shapes';

export class ExampleRecord extends Record({
  title: '',
  content: '',
  link: '',
  count: 0,
}) {
  resolveType() {
    if (/^https?:\/\//.test(this.link)) {
      return 'http';
    }
    if (this.link) {
      return 'link';
    }
    return 'text';
  }
}

export const exampleBox = createRecordBox('example', ExampleRecord);
