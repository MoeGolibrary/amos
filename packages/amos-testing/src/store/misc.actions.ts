/*
 * @since 2022-01-05 10:03:21
 * @author junbao <junbao@moego.pet>
 *
 * keep atomic boxes
 */

import { action } from 'amos-core';
import { countBox } from './misc.boxes';
import { sleep } from '../utils';

export const addTwiceAsync = action(async (dispatch, select, value: number) => {
  await sleep(1);
  dispatch(countBox.add(value));
  dispatch(countBox.add(value));
});
