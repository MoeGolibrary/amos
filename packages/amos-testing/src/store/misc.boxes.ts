/*
 * @since 2022-01-05 10:03:21
 * @author junbao <junbao@moego.pet>
 *
 * keep atomic boxes
 */

import { boolBox, numberBox } from 'amos-boxes';
import { box } from 'amos-core';
import { logoutSignal } from './session.signals';

export const countBox = numberBox('count');
logoutSignal.subscribe((dispatch) => dispatch(countBox.setState()));

export const darkModeBox = boolBox('ui.darkMode').config({ persist: { version: 1 } });

export const configBox = box('system.config', {
  api: 'https://github.com/amos-project/amos',
  env: 'testing',
});
