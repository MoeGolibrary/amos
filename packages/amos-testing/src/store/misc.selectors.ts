/*
 * @since 2022-01-05 10:03:21
 * @author junbao <junbao@moego.pet>
 *
 * keep atomic boxes
 */

import { selector } from 'amos-core';
import { countBox } from './misc.boxes';

export const selectCount = selector((select) => select(countBox));
export const selectDoubleCount = selector((select) => select(countBox) * 2);
export const selectMultipleCount = selector((select, times: number) => select(countBox) * times);

export const selectDouble = selector((select, v: number) => v * 2);
export const selectFourfold = selector(
  (select, v: number) => {
    return select(selectDouble(v)) * 2;
  },
  {
    cache: true,
  },
);

export const selectEightTimesCount = selector(
  (select, plus: number = 0) => {
    const count = select(countBox);
    return select(selectDouble(count)) * select(selectFourfold(count)) + plus;
  },
  {
    cache: true,
  },
);
