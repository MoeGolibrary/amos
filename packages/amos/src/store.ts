/*
 * @since 2020-11-03 13:31:31
 * @author acrazing <joking.young@gmail.com>
 */

import { Box } from './box';
import { SelectorFactory } from './selector';
import {
  Dispatch,
  Dispatchable,
  Select,
  Selectable,
  Snapshot,
  Subscribe,
  Unsubscribe,
} from './types';
import { CheapMap, CheapSet, isArray, threw, toString, warning } from './utils';

export interface StoreOptions {}

export interface SelectorCache<A extends any[] = any, R = any> {
  id: string;
  factory: SelectorFactory<A, R>;
  args: A;
  owners: CheapSet;
  usedBoxes: CheapSet;
  usedSelectors: CheapSet;
  value: R;
}

const VALUE_PLACEHOLDER = { '@AMOS_VALUE_PLACEHOLDER@': true };

export class Store {
  protected preloadedState: Snapshot;
  protected options: StoreOptions;
  protected listeners: Subscribe[];
  protected dispatching: Dispatchable | readonly Dispatchable[] | undefined;
  protected state: Snapshot;
  protected boxes: CheapMap<Box>;
  protected expiredBoxes: CheapSet;
  protected usedBoxes: CheapMap<CheapSet>;
  protected selectorStack: SelectorCache[];

  /**
   * The caches of selectors, please note it may be dirty.
   * @protected
   */
  protected selectorCaches: CheapMap<SelectorCache>;

  constructor(preloadedState: Snapshot = {}, options: StoreOptions = {}) {
    this.preloadedState = preloadedState;
    this.options = options;
    this.state = {};
    this.boxes = new CheapMap<Box>();
    this.listeners = [];
    this.dispatching = void 0;
    this.selectorStack = [];
    this.expiredBoxes = new CheapSet<string>();
    this.usedBoxes = new CheapMap<CheapSet>();
    this.selectorCaches = new CheapMap<SelectorCache>();
  }

  /**
   * executed after construct, used for enhancers.
   */
  init() {}

  /**
   * Take the snapshot of the state.
   *
   * Please note it is mutable.
   */
  snapshot(): Snapshot {
    return this.state;
  }

  /**
   * Subscribe the mutations.
   *
   * The handler will be notified when `store.dispatch` is called with state mutation.
   *
   * @param handler
   */
  subscribe(handler: Subscribe): Unsubscribe {
    this.listeners.push(handler);
    return () => {
      const index = this.listeners.indexOf(handler);
      if (index > -1) {
        this.listeners.splice(index, 1);
      }
    };
  }

  /**
   * Initialize a box, register it and preload its preloaded state.
   *
   * @param box
   * @protected
   */
  protected ensureBox(box: Box) {
    if (!this.boxes.hasOwnProperty(box.key)) {
      this.boxes.set(box.key, box);
    }
    if (this.state.hasOwnProperty(box.key)) {
      return this.state[box.key];
    }
    let boxState = box.initialState;
    if (this.preloadedState.hasOwnProperty(box.key)) {
      boxState = box.options.preload!(this.preloadedState[box.key], boxState);
    }
    this.state[box.key] = boxState;
    return boxState;
  }

  /**
   * Reset a box's state, notify the subscribers to re-initialize it.
   *
   * @param key
   * @protected
   */
  protected resetBox(key: string) {
    if (!this.boxes.hasOwnProperty(key)) {
      return;
    }
    delete this.state[key];
    this.ensureBox(this.boxes.take(key));
    this.notify();
  }

  /**
   * Execute a single dispatchable object.
   *
   * @param dispatchable
   * @protected
   */
  protected _exec(dispatchable: Dispatchable) {
    switch (dispatchable.$object) {
      case 'ACTION':
        // TODO: consider wrap dispatch with some context information
        //  to track the dispatch stack.
        return dispatchable.factory.actor(this.dispatch, this.select, ...dispatchable.args);
      case 'MUTATION':
        this.ensureBox(dispatchable.factory.box);
        const key = dispatchable.factory.box.key;
        const nextState = dispatchable.factory.mutator(this.state[key], ...dispatchable.args);
        if (nextState !== this.state[key]) {
          this.state[key] = nextState;
          this.expiredBoxes.add(key);
        }
        return nextState;
      case 'SIGNAL':
        for (const key in this.boxes) {
          if (!this.boxes.hasOwnProperty(key)) {
            continue;
          }
          const box = this.boxes.take(key);
          const fn = box.signals[dispatchable.type];
          if (!fn) {
            continue;
          }
          const nextState = fn(this.state[key], dispatchable.data);
          if (nextState !== this.state[key]) {
            this.state[key] = nextState;
            this.expiredBoxes.add(key);
          }
        }
        return dispatchable.data;
      default:
        threw(true, `Dispatch an object: ${dispatchable['$object'] || toString(dispatchable)}.`);
    }
  }

  /**
   * expose {@see _exec} to callbacks.
   *
   * @param dispatchable
   */
  protected exec = (dispatchable: Dispatchable) => this._exec(dispatchable);

  /**
   * Dispatch a dispatchable object or an array.
   *
   * If it is a transaction entry, notify the subscribers.
   *
   * @param tasks
   * @protected
   */
  protected _dispatch(tasks: Dispatchable | readonly Dispatchable[]) {
    if (process.env.NODE_ENV === 'development') {
      threw(!!this.selectorStack.length, 'You cannot dispatch a dispatchable in selector.');
    }
    const isRoot = !this.dispatching;
    if (isRoot) {
      this.dispatching = tasks;
    }
    try {
      if (isArray(tasks)) {
        return tasks.map(this.exec);
      } else {
        return this.exec(tasks);
      }
    } finally {
      if (isRoot) {
        this.dispatching = void 0;
        this.notify();
      }
    }
  }

  /**
   * Drop the expired selector caches.
   *
   * @protected
   */
  protected dropExpiredCaches() {
    const verified = new CheapSet();
    const drop = (id: string) => {};
    const verify = (id: string): boolean => {
      if (verified.has(id)) {
        // verified selectors, return directly.
        return this.selectorCaches.has(id);
      }
      const cache = this.selectorCaches.get(id);
      if (!cache) {
        // it is dropped already.
        return false;
      }
      if (cache.usedSelectors.every(verify)) {
        // every deps selector is ok, check boxes
        if (cache.usedBoxes.every((key) => !this.expiredBoxes.has(key))) {
          // box is also fine, it is fine
          verified.add(id);
          return true;
        } else {
          verified.add(id);
          if (cache.factory.expire === 'passive') {
            drop(id);
            return false;
          } else {
            try {
              const newValue = cache.factory.compute(this.select, ...cache.args);
              if (cache.factory.equal!(cache.value, newValue)) {
                return true;
              } else {
                drop(id);
                return false;
              }
            } catch (e) {
              warning(true, e?.stack || e + '');
              drop(id);
              return false;
            }
          }
        }
      } else {
        // if deps is expired, it will recompute all its dependents,
        // if expired, it should be dropped and added to verified already.
        return this.selectorCaches.has(id);
      }
    };

    this.expiredBoxes.forEach((key) => {
      const ref = this.usedBoxes.get(key);
      if (!ref) {
        return;
      }
      ref.forEach(verify);
    });
    this.expiredBoxes.clear();
  }

  /**
   * Notify the subscribers.
   *
   * @protected
   */
  protected notify() {
    this.listeners.forEach((fn) => fn());
  }

  /**
   * Expose {@see _dispatch} publicly as a property for callbacks.
   *
   * @param tasks
   */
  readonly dispatch: Dispatch = (tasks: any) => this._dispatch(tasks);

  /**
   * Select a selectable, and cache its result.
   *
   * A selectable could be:
   *
   * 1. a pure function, it will not be cached
   * 2. a selector instance
   * 3. a selector factory, it will not receive any args
   *
   * @param selectable
   * @protected
   */
  protected _select(selectable: Selectable) {
    this.dropExpiredCaches();
    // select function
    if (typeof selectable === 'function' && !('$object' in selectable)) {
      return selectable(this.select);
    }

    const parent = this.selectorStack[this.selectorStack.length - 1];

    // select box
    if (selectable instanceof Box) {
      const boxState = this.ensureBox(selectable);
      if (parent) {
        if (!this.usedBoxes.has(selectable.key)) {
          this.usedBoxes.set(selectable.key, new CheapSet([selectable.key]));
        } else {
          this.usedBoxes.take(selectable.key).add(selectable.key);
        }
        parent.usedBoxes.add(selectable.key);
      }
      return boxState;
    }

    // select selector or selector factory
    const args = selectable.$object === 'SELECTOR_FACTORY' ? [] : selectable.args;
    const factory = selectable.$object === 'SELECTOR_FACTORY' ? selectable : selectable.factory;
    try {
      const id = `${factory.type}${factory.id}.${factory.key!(this.select, ...args)}`;
      let node = this.selectorCaches.get(id);
      if (!node) {
        node = {
          id: id,
          factory: factory,
          args: args,
          owners: new CheapSet(),
          usedBoxes: new CheapSet(),
          usedSelectors: new CheapSet(),
          value: VALUE_PLACEHOLDER,
        };
      }
      this.selectorStack.push(node);
      if (parent) {
        node.owners.add(parent.id);
        parent.usedSelectors.add(id);
      }
      if (node.value === VALUE_PLACEHOLDER) {
        node.value = factory.compute(this.select, ...args);
      }
      return node.value;
    } finally {
      this.selectorStack.pop();
    }
  }

  /**
   * expose {@see _select} publicly and callbacks.
   *
   * @param selectable
   */
  select: Select = (selectable: Selectable) => this._select(selectable);

  /**
   * Reset internal state, used for HMR.
   *
   * The works includes:
   *
   * 1. Remove registered boxes
   * 2. Clean up selector caches
   * 3. Serialize and clean up state (works only if set serializeState).
   *
   * @param serializeState - if true, will serialize state and reload it with new boxes, you
   *                         should set is only if your box's state shape (include prototype)
   *                         is changed.
   */
  resetInternalState(serializeState: boolean) {
    if (process.env.NODE_ENV === 'development') {
      threw(
        !!this.dispatching || !!this.selectorStack.length,
        'You cannot call resetInternalState in an action or selector.',
      );
    }
    if (serializeState) {
      this.preloadedState = Object.assign(
        this.preloadedState || {},
        JSON.parse(JSON.stringify(this.state)),
      );
      this.state = {};
    }
    this.boxes.clear();
    this.expiredBoxes.clear();
    this.usedBoxes.clear();
    this.selectorCaches.clear();
    this.notify();
  }
}
