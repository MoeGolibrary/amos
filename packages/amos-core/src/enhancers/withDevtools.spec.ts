/*
 * @since 2024-10-25 11:25:04
 * @author junbao <junbao@moego.pet>
 */

import {
  addFourfoldAsync,
  countBox,
  expectCalled,
  expectCalledWith,
  loginAsync,
  LOGOUT,
  type LogoutEvent,
  sessionIdBox,
} from 'amos-testing';
import { createStore } from '../store';
import type { ReduxDevtoolsExtension } from './withDevtools';

describe('withDevtools', () => {
  it('should send event', async () => {
    const init = jest.fn();
    const send = jest.fn();
    const extension: ReduxDevtoolsExtension = {
      connect: () => ({ init, send }),
    };
    const e: LogoutEvent = { userId: 2, sessionId: 2 };
    const store = createStore({ devtools: { enable: true, extension } });
    const tasks = [addFourfoldAsync(2), loginAsync(1), LOGOUT(e), countBox.setState(1)] as const;
    const r = await Promise.all(store.dispatch(tasks));
    expect([...r, store.select(countBox)]).toEqual([8, store.select(sessionIdBox), e, 1, 5]);
    expectCalledWith(init, [{}]);
    const calls = Object.fromEntries(
      send.mock.calls.map((c, i, array) => [
        i.toString().padStart(array.length.toString().length, '0') + ':' + c[0].type,
        {
          ...c[0],
          root: c[0].root && 'ROOT',
          state: c[1],
        },
      ]),
    );
    expect(calls).toMatchSnapshot('devtools events');
    expectCalled(send, 21);
  });
});
