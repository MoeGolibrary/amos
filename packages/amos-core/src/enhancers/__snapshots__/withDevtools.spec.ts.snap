// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDevtools should send event: devtools events 1`] = `
{
  "00:P:count": {
    "args": [
      0,
    ],
    "root": "ROOT",
    "state": {
      "count": 0,
    },
    "type": "P:count",
  },
  "01:M:count/add": {
    "args": [
      2,
    ],
    "root": "ROOT",
    "state": {
      "count": 2,
    },
    "type": "M:count/add",
  },
  "02:M:count/setState": {
    "args": [
      [Function],
    ],
    "root": "ROOT",
    "state": {
      "count": 4,
    },
    "type": "M:count/setState",
  },
  "03:A:amos/addFourfoldAsync": {
    "args": [
      2,
    ],
    "root": "ROOT",
    "state": {
      "count": 4,
    },
    "type": "A:amos/addFourfoldAsync",
  },
  "04:A:amos/loginAsync": {
    "args": [
      1,
    ],
    "root": "ROOT",
    "state": {
      "count": 4,
    },
    "type": "A:amos/loginAsync",
  },
  "05:M:count/setState": {
    "args": [],
    "root": "ROOT",
    "state": {
      "count": 0,
    },
    "type": "M:count/setState",
  },
  "06:P:sessions": {
    "args": [
      {},
    ],
    "root": "ROOT",
    "state": {
      "count": 0,
      "sessions": {},
    },
    "type": "P:sessions",
  },
  "07:M:sessions/setState": {
    "args": [
      {},
    ],
    "root": "ROOT",
    "state": {
      "count": 0,
      "sessions": {},
    },
    "type": "M:sessions/setState",
  },
  "08:S:session.logout": {
    "args": [
      {
        "sessionId": 2,
        "userId": 2,
      },
    ],
    "root": "ROOT",
    "state": {
      "count": 0,
      "sessions": {},
    },
    "type": "S:session.logout",
  },
  "09:M:count/setState": {
    "args": [
      1,
    ],
    "root": "ROOT",
    "state": {
      "count": 1,
      "sessions": {},
    },
    "type": "M:count/setState",
  },
  "10:A:amos/batch": {
    "args": [
      [
        {
          "actor": [Function],
          "args": [
            2,
          ],
          "conflictPolicy": "always",
          "id": "T-7",
          "type": "amos/addFourfoldAsync",
          Symbol($amos): "action",
        },
        {
          "actor": [Function],
          "args": [
            1,
          ],
          "conflictPolicy": "always",
          "id": "T-44",
          "type": "amos/loginAsync",
          Symbol($amos): "action",
        },
        {
          "args": [
            {
              "sessionId": 2,
              "userId": 2,
            },
          ],
          "creator": [Function],
          "factory": [Function],
          "id": "T-54",
          "type": "session.logout",
          Symbol($amos): "signal",
        },
        {
          "args": [
            1,
          ],
          "box": NumberBox {
            "id": "T-3",
            "initialState": 0,
            "key": "count",
            Symbol($amos): "box",
          },
          "id": "T-3/setState",
          "mutator": [Function],
          "type": "count/setState",
          Symbol($amos): "mutation",
        },
      ],
    ],
    "root": undefined,
    "state": {
      "count": 1,
      "sessions": {},
    },
    "type": "A:amos/batch",
  },
  "11:M:count/setState": {
    "args": [
      3,
    ],
    "root": undefined,
    "state": {
      "count": 3,
      "sessions": {},
    },
    "type": "M:count/setState",
  },
  "12:M:count/add": {
    "args": [
      2,
    ],
    "root": undefined,
    "state": {
      "count": 5,
      "sessions": {},
    },
    "type": "M:count/add",
  },
  "13:M:sessions/mergeItem": {
    "args": [
      10001,
      {
        "id": 10001,
        "userId": 1,
      },
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
    },
    "type": "M:sessions/mergeItem",
  },
  "14:P:sessions.currentId": {
    "args": [
      0,
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 0,
    },
    "type": "P:sessions.currentId",
  },
  "15:M:sessions.currentId/setState": {
    "args": [
      10001,
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
    },
    "type": "M:sessions.currentId/setState",
  },
  "16:P:user.userMap": {
    "args": [
      {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
      "user.userMap": {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    },
    "type": "P:user.userMap",
  },
  "17:M:user.userMap/mergeItem": {
    "args": [
      1,
      {
        "id": 1,
      },
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
      "user.userMap": {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    },
    "type": "M:user.userMap/mergeItem",
  },
  "18:A:amos/batch": {
    "args": [
      [
        {
          "args": [
            10001,
            {
              "id": 10001,
              "userId": 1,
            },
          ],
          "box": RecordMapBox {
            "id": "T-30",
            "initialState": {},
            "key": "sessions",
            Symbol($amos): "box",
          },
          "id": "T-30/mergeItem",
          "mutator": [Function],
          "type": "sessions/mergeItem",
          Symbol($amos): "mutation",
        },
        {
          "args": [
            10001,
          ],
          "box": Box {
            "id": "T-31",
            "initialState": 0,
            "key": "sessions.currentId",
            Symbol($amos): "box",
          },
          "id": "T-31/setState",
          "mutator": [Function],
          "type": "sessions.currentId/setState",
          Symbol($amos): "mutation",
        },
        {
          "args": [
            1,
            {
              "id": 1,
            },
          ],
          "box": RecordMapBox {
            "id": "T-51",
            "initialState": {
              "1": {
                "createdAt": 2020-01-01T08:00:00.000Z,
                "fatherId": 0,
                "firstName": "Rick",
                "id": 1,
                "lastName": "Sanchez",
                "motherId": 0,
              },
              "2": {
                "createdAt": 2020-01-01T08:00:00.000Z,
                "fatherId": 1,
                "firstName": "Beth",
                "id": 2,
                "lastName": "Sanchez",
                "motherId": 0,
              },
              "3": {
                "createdAt": 2020-01-01T08:00:00.000Z,
                "fatherId": 0,
                "firstName": "Jerry",
                "id": 3,
                "lastName": "Sanchez",
                "motherId": 0,
              },
              "4": {
                "createdAt": 2020-01-01T08:00:00.000Z,
                "fatherId": 3,
                "firstName": "Jessica",
                "id": 4,
                "lastName": "Sanchez",
                "motherId": 0,
              },
              "5": {
                "createdAt": 2020-01-01T08:00:00.000Z,
                "fatherId": 0,
                "firstName": "Morty",
                "id": 5,
                "lastName": "Sanchez",
                "motherId": 0,
              },
            },
            "key": "user.userMap",
            Symbol($amos): "box",
          },
          "id": "T-51/mergeItem",
          "mutator": [Function],
          "type": "user.userMap/mergeItem",
          Symbol($amos): "mutation",
        },
      ],
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
      "user.userMap": {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    },
    "type": "A:amos/batch",
  },
  "19:S:session.login": {
    "args": [
      {
        "sessionId": 10001,
        "userId": 1,
      },
    ],
    "root": "ROOT",
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
      "user.userMap": {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    },
    "type": "S:session.login",
  },
  "20:A:amos/loginSync": {
    "args": [
      1,
    ],
    "root": undefined,
    "state": {
      "count": 5,
      "sessions": {
        "10001": {
          "id": 10001,
          "token": "",
          "userId": 1,
        },
      },
      "sessions.currentId": 10001,
      "user.userMap": {
        "1": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Rick",
          "id": 1,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "2": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 1,
          "firstName": "Beth",
          "id": 2,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "3": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Jerry",
          "id": 3,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "4": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 3,
          "firstName": "Jessica",
          "id": 4,
          "lastName": "Sanchez",
          "motherId": 0,
        },
        "5": {
          "createdAt": 2020-01-01T08:00:00.000Z,
          "fatherId": 0,
          "firstName": "Morty",
          "id": 5,
          "lastName": "Sanchez",
          "motherId": 0,
        },
      },
    },
    "type": "A:amos/loginSync",
  },
}
`;
